core/auth/src/test/java/software/amazon/awssdk/auth/credentials/DefaultCredentialsProviderTest.java

```diff
@@ -97,6 +107,44 @@ void resolveCredentials_ProfileCredentialsProviderWithSupplierProfileFile_return
         });
     }
 
+    @Test
+    void resolveCredentials_DefaultCredentialProviderWithReloadWhenModified() throws Exception {
+        Path credentialsFilePath = generateTestCredentialsFile("customAccess", "customSecret");
+        SystemSettingUtilsTestBackdoor.addEnvironmentVariableOverride(ProfileFileSystemSetting.AWS_SHARED_CREDENTIALS_FILE.environmentVariable(),
+                                                                      credentialsFilePath.toString());
+        DefaultCredentialsProvider provider = DefaultCredentialsProvider.create();
+
+        assertThat(provider.resolveCredentials()).satisfies(awsCredentials -> {
+            assertThat(awsCredentials.accessKeyId()).isEqualTo("customAccess");
+            assertThat(awsCredentials.secretAccessKey()).isEqualTo("customSecret");
+        });
+
+        generateTestCredentialsFile("modifiedAccess", "modifiedSecret");
+
+        // without sleep the assertion fails, as there is delay in config reload
+        Thread.sleep(2000);
```

Comment:
@anirudh9391 For Java8, the test is failing pos this, had included delay to ensure latest config is picked up, but it is failing occasionally, 
```
Error:    DefaultCredentialsProviderTest.resolveCredentials_DefaultCredentialProviderWithReloadWhenModified:126->lambda$resolveCredentials_DefaultCredentialProviderWithReloadWhenModified$7:127
expected: "modifiedAccess"
 but was: "customAccess"

Should we increase the sleep value? Is there any alternative way to validate this?
```
For Java17, the test failures seems to be unrelated & from s3 module 
```
2024-06-11T16:42:29.7804890Z [ERROR] Errors:
2024-06-11T16:42:29.7805324Z [ERROR]   CrtDownloadErrorTest.verifyCrtResource:73 ? Runtime Timeout waiting for resource count to drop to zero
2024-06-11T16:42:29.7805829Z [ERROR]   CrtCredentialProviderAdapterTest.verifyCrtResource:43 ? Runtime Timeout waiting for resource count to drop to zero
2024-06-11T16:42:29.7806270Z [ERROR]   S3CrtClientWiremockTest.verifyCrtResource:92 ? Runtime Timeout waiting for resource count to drop to zero
```

Reply:
The Java17 error seems to be transient in nature. It seems to succeed locally. I will play around with timeout values a bit.

Reply:
Due to the caching of ProfileFile for short duration, https://github.com/aws/aws-sdk-java-v2/blob/fb7b6ea57c5512fa6229af06b904070054ed1597/utils/src/main/java/software/amazon/awssdk/utils/cache/CachedSupplier.java#L242
the credentials are not reloaded immediately, hence, Thread.sleep is required, added 2s sleep which in most cases sufficient, is not working in CI sometimes. 
Here, sleep duration can be increased or assertion can retried with smaller timeout values

For reference, in ProfileRefresherTest https://github.com/aws/aws-sdk-java-v2/blob/fb7b6ea57c5512fa6229af06b904070054ed1597/core/profiles/src/test/java/software/amazon/awssdk/profiles/internal/ProfileFileRefresherTest.java#L134 

Custom clock is used to advance the timer to validate refresh without sleep which might not be possible here, ie to set custom clock in ProfileSupplier

Reply:
I tried it with a 5 second delay. Seems to trip up the java21 and java8 builds, both of which fail assertion. Please feel free to cut another revision with any modifications to add retries or anything else. I think we can keep timeout at 5 seconds.

Reply:
I pushed another revision with a retry. Lets see if this works

Reply:
Seems like the retry failed all the java builds

Reply:
Thanks @anirudh9391 , will take a look. Going through the commit, I think retry should be added to second assertion, once credentials are modified

Reply:
Updated the retry logic, along with additional logging to debug the test, if it fails again in the commits https://github.com/aws/aws-sdk-java-v2/pull/5185/commits/e360e9d13e1e5485110306ad27fc4f02619211c9 and https://github.com/aws/aws-sdk-java-v2/pull/5185/commits/f3304575a2b4315d01e666a50edd837abbcf472c respectively

Reply:
@anirudh9391 could you please cherry-pick latest commit for the fix, and further debugging?